File: src/test/unit/algorithms/SDRClassifierTest.cpp
/* ---------------------------------------------------------------------
 * HTM Community Edition of NuPIC
 * Copyright (C) 2016, Numenta, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero Public License for more details.
 *
 * You should have received a copy of the GNU Affero Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 * --------------------------------------------------------------------- */

/** @file
 * Implementation of unit tests for Classifier & Predictor
 */

#include <cmath> // isnan
#include <iostream>
#include <limits> // numeric_limits
#include <sstream>
#include <stdio.h>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include <htm/algorithms/SDRClassifier.hpp>
#include <htm/utils/Log.hpp>

using namespace std;
using namespace htm;

namespace testing {


TEST(SDRClassifierTest, ExampleUsageClassifier)
{
  // Make a random SDR and associate it with the category B.
  SDR inputData({ 1000u });
      inputData.randomize( 0.02f );
  enum Category { A, B, C, D };
  Classifier clsr;
  clsr.learn( inputData, { Category::B } );
  ASSERT_EQ( argmax( clsr.infer( inputData ) ),  Category::B );

  // Estimate a scalar value.  The Classifier only accepts categories, so
  // put real valued inputs into bins (AKA buckets) by subtracting the
  // minimum value and dividing by a resolution.
  double scalar = 567.8f;
  double minimum = 500.0f;
  double resolution = 10.0f;
  clsr.learn( inputData, { (UInt)((scalar - minimum) / resolution) } );
  ASSERT_EQ( argmax( clsr.infer( inputData ) ) * resolution + minimum,  560.0f );
}


TEST(SDRClassifierTest, ExampleUsagePredictor)
{
  // Predict 1 and 2 time steps into the future.

  // Make a sequence of 4 random SDRs. Each SDR has 1000 bits and 2% sparsity.
  vector<SDR> sequence( 4u, vector<UInt>{ 1000u } );
  for( SDR & inputData : sequence ) {
      inputData.randomize( 0.02f );
  }
  // Make category labels for the sequence.
  vector<UInt> labels = { 4, 5, 6, 7 };

  // Make a Predictor and train it.
  Predictor pred( vector<UInt>{ 1, 2 } );
  pred.learn( 0, sequence[0], { labels[0] } );
  pred.learn( 1, sequence[1], { labels[1] } );
  pred.learn( 2, sequence[2], { labels[2] } );
  pred.learn( 3, sequence[3], { labels[3] } );

  // Give the predictor partial information, and make predictions
  // about the future.
  pred.reset();
  Predictions A = pred.infer( sequence[0] );
  ASSERT_EQ( argmax( A[1] ),  labels[1] );
  ASSERT_EQ( argmax( A[2] ),  labels[2] );

  Predictions B = pred.infer( sequence[1] );
  ASSERT_EQ( argmax( B[1] ),  labels[2] );
  ASSERT_EQ( argmax( B[2] ),  labels[3] );
}


TEST(SDRClassifierTest, SingleValue) {
  // Feed the same input 10 times, the corresponding probability should be
  // very high
  vector<UInt> steps{1u};
  Predictor c(steps, 0.1f);

  // Create a vector of input bit indices
  SDR input1({10u}); input1.setSparse(SDR_sparse_t({ 1u, 5u, 9u }));
  vector<UInt> bucketIdxList{4u};
  for (UInt i = 0u; i < 10u; ++i) {
    c.learn( i, input1, bucketIdxList );
  }
  Predictions result1 = c.infer( input1 );

  ASSERT_EQ( argmax( result1[1u] ), 4u )
      << "Incorrect prediction for bucket 4";

  ASSERT_EQ( result1.size(), 1u );
}


TEST(SDRClassifierTest, ComputeComplex) {
  // More complex classification
  // This test is ported from the Python unit test
  Predictor c({1u}, 1.0f);

  // Create a input vector
  SDR input1({ 20u });
  input1.setSparse(SDR_sparse_t({ 1u, 5u, 9u }));
  vector<UInt> bucketIdxList1{ 4u };

  // Create a input vector
  SDR input2({ 20u });
  input2.setSparse(SDR_sparse_t({ 0u, 6u, 9u, 11u }));
  vector<UInt> bucketIdxList2{ 5u };

  // Create input vectors
  SDR input3({ 20u });
  input3.setSparse(SDR_sparse_t({ 6u, 9u }));
  vector<UInt> bucketIdxList3{ 5u };
  vector<UInt> bucketIdxList4{ 4u };
  vector<UInt> bucketIdxList5{ 4u };

  c.learn(0, input1, bucketIdxList1);
  c.learn(1, input2, bucketIdxList2);
  c.learn(2, input3, bucketIdxList3);
  c.learn(3, input1, bucketIdxList4);
  auto result = c.infer(input1);

  // Check the one-step prediction
  ASSERT_EQ(result.size(), 1u)
    << "Result should only have 1 key.";
  ASSERT_EQ(6ul, result[1u].size()) << "Expected six bucket predictions";
  ASSERT_LT(fabs(result[1u].at(0u) - 0.034234f), 0.000001f)
      << "Incorrect prediction for bucket 0";
  ASSERT_LT(fabs(result[1u].at(1u) - 0.034234f), 0.000001f)
      << "Incorrect prediction for bucket 1";
  ASSERT_LT(fabs(result[1u].at(2u) - 0.034234f), 0.000001f)
      << "Incorrect prediction for bucket 2";
  ASSERT_LT(fabs(result[1u].at(3u) - 0.034234f), 0.000001f)
      << "Incorrect prediction for bucket 3";
  ASSERT_LT(fabs(result[1u].at(4u) - 0.093058f), 0.000001f)
      << "Incorrect prediction for bucket 4";
  ASSERT_LT(fabs(result[1u].at(5u) - 0.770004f), 0.000001f)
      << "Incorrect prediction for bucket 5";
}


TEST(SDRClassifierTest, MultipleCategories) {
  // Test multiple category classification with single compute calls
  // This test is ported from the Python unit test
  Classifier c(1.0f);

  // Create a input vectors
  SDR input1({ 10 });
  input1.setSparse(SDR_sparse_t({ 1u, 3u, 5u }));
  vector<UInt> bucketIdxList1{ 0u, 1u };

  // Create a input vectors
  SDR input2({ 10 });
  input2.setSparse(SDR_sparse_t({ 2u, 4u, 6u }));
  vector<UInt> bucketIdxList2{ 2u, 3u };

  // Train
  for (auto i = 0u; i < 1000u; i++) {
    c.learn( input1, bucketIdxList1 );
    c.learn( input2, bucketIdxList2 );
  }

  // Test
  PDF result1 = c.infer( input1 );
  PDF result2 = c.infer( input2 );

  ASSERT_LT(fabs(result1.at(0u) - 0.5f), 0.1f)
      << "Incorrect prediction for bucket 0 (expected=0.5)";
  ASSERT_LT(fabs(result1.at(1u) - 0.5f), 0.1f)
      << "Incorrect prediction for bucket 1 (expected=0.5)";

  ASSERT_LT(fabs(result2.at(2u) - 0.5f), 0.1f)
      << "Incorrect prediction for bucket 2 (expected=0.5)";
  ASSERT_LT(fabs(result2.at(3u) - 0.5f), 0.1f)
      << "Incorrect prediction for bucket 3 (expected=0.5)";
}


TEST(SDRClassifierTest, SaveLoad) {
  vector<UInt> steps{ 1u };
  Predictor c1(steps, 0.1f);

  // Train a Predictor with a few different things.
  SDR A({ 100u }); A.randomize( 0.10f );
  for(UInt i = 0; i < 10u; i++)
    { c1.learn(i, A, {4u}); }
  c1.reset();
  A.addNoise( 1.0f ); // Change every bit.
  for(UInt i = 0; i < 10u; i++)
    { c1.learn(i, A, {3u, 5u}); }
  // Measure and save some output.
  A.addNoise( 0.20f ); // Change two bits.
  c1.reset();
  const auto c1_out = c1.infer( A );

  // Save and load.
  stringstream ss;
  EXPECT_NO_THROW(c1.save(ss));
  Predictor c2;
  EXPECT_NO_THROW(c2.load(ss));

  // Expect identical results.
  const auto c2_out = c2.infer( A );
  ASSERT_EQ(c1_out, c2_out);
}


TEST(SDRClassifierTest, testSoftmaxOverflow) {
  PDF values({ numeric_limits<Real>::max() });
  softmax(values.begin(), values.end());
  auto result = values[0u];
  ASSERT_FALSE(std::isnan(result));
}


TEST(SDRClassifierTest, testSoftmax) {
  PDF values {0.0f, 1.0f, 1.337f, 2.018f, 1.1f, 0.5f, 0.9f};
  const PDF exp {
    0.045123016137150938f,
    0.12265707481088166f,
    0.17181055613150184f,
    0.3394723335640627f,
    0.13555703197721547f,
    0.074395276503465876f,
    0.11098471087572169f};

  softmax(values.begin(), values.end());

  for(auto i = 0u; i < exp.size(); i++) {
    EXPECT_NEAR(values[i], exp[i], 0.000001f) << "softmax ["<< i <<"]";
  }
}

} // end namespace

File: bindings/py/tests/algorithms/sdr_classifier_test.py
# ----------------------------------------------------------------------
# Numenta Platform for Intelligent Computing (NuPIC)
# Copyright (C) 2016, Numenta, Inc.
#               2019, David McDougall
#
# Unless you have an agreement with Numenta, Inc., for a separate license for
# this software code, the following terms and conditions apply:
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Affero Public License for more details.
#
# You should have received a copy of the GNU Affero Public License
# along with this program.  If not, see http://www.gnu.org/licenses.
#
# http://numenta.org/licenses/
# ----------------------------------------------------------------------

""" Unit tests for Classifier & Predictor classes. """

import numpy
import pickle
import random
import sys
import tempfile
import unittest

from nupic.bindings.sdr import SDR
from nupic.bindings.algorithms import Classifier, Predictor


class ClassifierTest(unittest.TestCase):
  """ Unit tests for Classifier & Predictor classes. """


  def testExampleUsage(self):
    # Make a random SDR and associate it with a category.
    inputData  = SDR( 1000 ).randomize( 0.02 )
    categories = { 'A': 0, 'B': 1, 'C': 2, 'D': 3 }
    clsr = Classifier()
    clsr.learn( inputData, categories['B'] )
    assert( numpy.argmax( clsr.infer( inputData ) )  ==  categories['B'] )

    # Estimate a scalar value.  The Classifier only accepts categories, so
    # put real valued inputs into bins (AKA buckets) by subtracting the
    # minimum value and dividing by a resolution.
    scalar     = 567.8
    minimum    = 500
    resolution = 10
    clsr.learn( inputData, int((scalar - minimum) / resolution) )
    assert( numpy.argmax( clsr.infer( inputData ) ) * resolution + minimum  ==  560 )


    # Predict 1 and 2 time steps into the future.

    # Make a sequence of 4 random SDRs, each SDR has 1000 bits and 2% sparsity.
    sequence = [ SDR( 1000 ).randomize( 0.02 ) for i in range(4) ]

    # Make category labels for the sequence.
    labels = [ 4, 5, 6, 7 ]

    # Make a Predictor and train it.
    pred = Predictor([ 1, 2 ])
    pred.learn( 0, sequence[0], labels[0] )
    pred.learn( 1, sequence[1], labels[1] )
    pred.learn( 2, sequence[2], labels[2] )
    pred.learn( 3, sequence[3], labels[3] )

    # Give the predictor partial information, and make predictions
    # about the future.
    pred.reset()
    A = pred.infer( 0, sequence[0] )
    assert( numpy.argmax( A[1] )  ==  labels[1] )
    assert( numpy.argmax( A[2] )  ==  labels[2] )

    B = pred.infer( 1, sequence[1] )
    assert( numpy.argmax( B[1] )  ==  labels[2] )
    assert( numpy.argmax( B[2] )  ==  labels[3] )


  def testInitialization(self):
    Classifier( .1 )
    Predictor( [2,3,4], .1 )


  def testInitInvalidParams(self):
    # Invalid alpha
    self.assertRaises(Exception, Classifier, alpha = -1.0)

    # Invalid steps
    kwargs = {"steps": [0.3], "alpha": 0.1}
    self.assertRaises(Exception, Predictor, **kwargs)
    kwargs = {"steps": [], "alpha": 0.1}
    self.assertRaises(Exception, Predictor, **kwargs)
    kwargs = {"steps": [-1], "alpha": 0.1}
    self.assertRaises(Exception, Predictor, **kwargs)


  def testSingleValue(self):
    """Send same value 10 times and expect high likelihood for prediction."""
    classifier = Classifier( alpha = 0.5 )

    # Enough times to perform inference and learn associations
    inp = SDR(10)
    inp.randomize( .2 )
    for recordNum in range(10):
      classifier.learn(inp, 2)

    retval = classifier.infer( inp )
    self.assertGreater(retval[2], 0.9)


  def testSingleValue0Steps(self):
    """Send same value 10 times and expect high likelihood for prediction
    using 0-step ahead prediction"""
    pred = Predictor( steps=[0], alpha=0.5)

    # Enough times to perform Inference and learn associations
    inp = SDR(10)
    inp.randomize( .2 )
    for recordNum in range(10):
      pred.learn(recordNum, inp, 2)

    retval = pred.infer( 10, inp )
    self.assertGreater(retval[0][2], 0.9)


  def testComputeInferOrLearnOnly(self):
    c = Predictor([1], 1.0)
    inp = SDR(10)
    inp.randomize( .3 )

    # learn only
    c.infer(recordNum=0, pattern=inp) # Don't crash with not enough training data.
    c.learn(recordNum=0, pattern=inp, classification=4)
    c.infer(recordNum=1, pattern=inp) # Don't crash with not enough training data.
    c.learn(recordNum=2, pattern=inp, classification=4)
    c.learn(recordNum=3, pattern=inp, classification=4)

    # infer only
    retval1 = c.infer(recordNum=5, pattern=inp)
    retval2 = c.infer(recordNum=6, pattern=inp)
    self.assertSequenceEqual(list(retval1[1]), list(retval2[1]))


  def testComputeComplex(self):
    c   = Predictor([1], 1.0)
    inp = SDR(100)

    inp.sparse = [1, 5, 9]
    c.learn(recordNum=0, pattern=inp,
              classification=4,)

    inp.sparse = [0, 6, 9, 11]
    c.learn(recordNum=1, pattern=inp,
              classification=5,)

    inp.sparse = [6, 9]
    c.learn(recordNum=2, pattern=inp,
              classification=5,)

    inp.sparse = [1, 5, 9]
    c.learn(recordNum=3, pattern=inp,
              classification=4,)

    inp.sparse = [1, 5, 9]
    result = c.infer(recordNum=4, pattern=inp)

    self.assertSetEqual(set(result.keys()), set([1]))
    self.assertEqual(len(result[1]), 6)
    self.assertAlmostEqual(result[1][0], 0.034234, places=5)
    self.assertAlmostEqual(result[1][1], 0.034234, places=5)
    self.assertAlmostEqual(result[1][2], 0.034234, places=5)
    self.assertAlmostEqual(result[1][3], 0.034234, places=5)
    self.assertAlmostEqual(result[1][4], 0.093058, places=5)
    self.assertAlmostEqual(result[1][5], 0.770004, places=5)


  def testOverlapPattern(self):
    classifier = Classifier(alpha=10.0)
    inp = SDR( 10 )
    inp.randomize( .2 )

    classifier.learn(pattern=inp, classification=9)
    classifier.learn(pattern=inp, classification=9)

    inp.addNoise( .5 )
    retval = classifier.infer(pattern=inp)

    # Since overlap - should be previous with high likelihood
    self.assertGreater(retval[9], 0.9)

    classifier.learn(pattern=inp, classification=2)
    classifier.learn(pattern=inp, classification=2)
    # Second example: now new value should be more probable than old

    retval = classifier.infer(pattern=inp)
    self.assertGreater(retval[2], retval[9])


  def testMultistepSingleValue(self):
    classifier = Predictor(steps=[1, 2])
    inp = SDR( 10 )
    inp.randomize( .2 )

    for recordNum in range(10):
      classifier.learn(recordNum, inp, 0)

    retval = classifier.infer(10, inp)

    # Should have a probability of 100% for that bucket.
    self.assertEqual(retval[1], [1.])
    self.assertEqual(retval[2], [1.])


  def testMultistepSimple(self):
    classifier = Predictor(steps=[1, 2], alpha=10.0)
    inp = SDR( 10 )

    for i in range(100):
      inp.sparse = [i % 10]
      classifier.learn(recordNum=i, pattern=inp, classification=(i % 10))

    retval = classifier.infer(99, inp)

    self.assertGreater(retval[1][0], 0.99)
    for i in range(1, 10):
      self.assertLess(retval[1][i], 0.01)
    self.assertGreater(retval[2][1], 0.99)
    for i in [0] + list(range(2, 10)):
      self.assertLess(retval[2][i], 0.01)


  def testMissingRecords(self):
    """ Test missing record support.

    Here, we intend the classifier to learn the associations:
      [1,3,5] => bucketIdx 1
      [2,4,6] => bucketIdx 2
      [7,8,9] => don"t care

    If it doesn't pay attention to the recordNums in this test, it will learn the
    wrong associations.
    """

    c = Predictor( steps=[1], alpha=1.0 )
    recordNum = 0
    inp = SDR( 10 )

    inp.sparse = [1, 3, 5]
    c.learn(recordNum=recordNum, pattern=inp, classification=0)
    recordNum += 1

    inp.sparse = [2, 4, 6]
    c.learn(recordNum=recordNum, pattern=inp, classification=1)
    recordNum += 1

    inp.sparse = [1, 3, 5]
    c.learn(recordNum=recordNum, pattern=inp, classification=2)
    recordNum += 1

    inp.sparse = [2, 4, 6]
    c.learn(recordNum=recordNum, pattern=inp, classification=1)
    recordNum += 1

    # -----------------------------------------------------------------------
    # At this point, we should have learned [1,3,5] => bucket 1
    #                                       [2,4,6] => bucket 2
    inp.sparse = [1, 3, 5]
    result = c.infer(recordNum=recordNum, pattern=inp)
    c.learn(recordNum=recordNum, pattern=inp, classification=2)
    recordNum += 1
    self.assertLess(result[1][0], 0.1)
    self.assertGreater(result[1][1], 0.9)
    self.assertLess(result[1][2], 0.1)

    inp.sparse = [2, 4, 6]
    result = c.infer(recordNum=recordNum, pattern=inp)
    c.learn(recordNum=recordNum, pattern=inp, classification=1)
    recordNum += 1
    self.assertLess(result[1][0], 0.1)
    self.assertLess(result[1][1], 0.1)
    self.assertGreater(result[1][2], 0.9)

    # -----------------------------------------------------------------------
    # Feed in records that skip and make sure they don"t mess up what we
    #  learned
    # If we skip a record, the CLA should NOT learn that [2,4,6] from
    #  the previous learn associates with bucket 0
    recordNum += 1
    inp.sparse = [1, 3, 5]
    result = c.infer(recordNum=recordNum, pattern=inp)
    c.learn(recordNum=recordNum, pattern=inp, classification=0)
    recordNum += 1
    self.assertLess(result[1][0], 0.1)
    self.assertGreater(result[1][1], 0.9)
    self.assertLess(result[1][2], 0.1)

    # If we skip a record, the CLA should NOT learn that [1,3,5] from
    #  the previous learn associates with bucket 0
    recordNum += 1
    inp.sparse = [2, 4, 6]
    result = c.infer(recordNum=recordNum, pattern=inp)
    c.learn(recordNum=recordNum, pattern=inp, classification=0)
    recordNum += 1
    self.assertLess(result[1][0], 0.1)
    self.assertLess(result[1][1], 0.1)
    self.assertGreater(result[1][2], 0.9)

    # If we skip a record, the CLA should NOT learn that [2,4,6] from
    #  the previous learn associates with bucket 0
    recordNum += 1
    inp.sparse = [1, 3, 5]
    result = c.infer(recordNum=recordNum, pattern=inp)
    c.learn(recordNum=recordNum, pattern=inp, classification=0)
    recordNum += 1
    self.assertLess(result[1][0], 0.1)
    self.assertGreater(result[1][1], 0.9)
    self.assertLess(result[1][2], 0.1)


  def testPredictionDistribution(self):
    """ Test the distribution of predictions.

    Here, we intend the classifier to learn the associations:
      [1,3,5] => bucketIdx 0 (30%)
              => bucketIdx 1 (30%)
              => bucketIdx 2 (40%)

      [2,4,6] => bucketIdx 1 (50%)
              => bucketIdx 3 (50%)

    The classifier should get the distribution almost right given enough
    repetitions and a small learning rate
    """

    c = Classifier(alpha = 0.001)

    SDR1 = SDR(10);  SDR1.sparse = [1, 3, 5]
    SDR2 = SDR(10);  SDR2.sparse = [2, 4, 6]

    random.seed(42)
    for _ in range(5000):
      randomNumber = random.random()
      if randomNumber < 0.3:
        bucketIdx = 0
      elif randomNumber < 0.6:
        bucketIdx = 1
      else:
        bucketIdx = 2
      c.learn(pattern=SDR1, classification=bucketIdx)

      randomNumber = random.random()
      if randomNumber < 0.5:
        bucketIdx = 1
      else:
        bucketIdx = 3
      c.learn(pattern=SDR2, classification=bucketIdx)

    result1 = c.infer(pattern=SDR1)
    self.assertAlmostEqual(result1[0], 0.3, places=1)
    self.assertAlmostEqual(result1[1], 0.3, places=1)
    self.assertAlmostEqual(result1[2], 0.4, places=1)

    result2 = c.infer(pattern=SDR2)
    self.assertAlmostEqual(result2[1], 0.5, places=1)
    self.assertAlmostEqual(result2[3], 0.5, places=1)


  def testPredictionDistributionOverlap(self):
    """ Test the distribution of predictions with overlapping input SDRs

    Here, we intend the classifier to learn the associations:
      SDR1    => bucketIdx 0 (30%)
              => bucketIdx 1 (30%)
              => bucketIdx 2 (40%)

      SDR2    => bucketIdx 1 (50%)
              => bucketIdx 3 (50%)

    SDR1 and SDR2 has 10% overlaps (2 bits out of 20)
    The classifier should get the distribution almost right despite the overlap
    """
    c = Classifier( 0.0005 )

    # generate 2 SDRs with 2 shared bits
    SDR1 = SDR( 100 )
    SDR2 = SDR( 100 )
    SDR1.randomize( .20 )
    SDR2.setSDR( SDR1 )
    SDR2.addNoise( .9 )

    random.seed(42)
    for _ in range(5000):
      randomNumber = random.random()
      if randomNumber < 0.3:
        bucketIdx = 0
      elif randomNumber < 0.6:
        bucketIdx = 1
      else:
        bucketIdx = 2
      c.learn(SDR1, bucketIdx)

      randomNumber = random.random()
      if randomNumber < 0.5:
        bucketIdx = 1
      else:
        bucketIdx = 3
      c.learn(SDR2, bucketIdx)

    result1 = c.infer(SDR1)
    self.assertAlmostEqual(result1[0], 0.3, places=1)
    self.assertAlmostEqual(result1[1], 0.3, places=1)
    self.assertAlmostEqual(result1[2], 0.4, places=1)

    result2 = c.infer(SDR2)
    self.assertAlmostEqual(result2[1], 0.5, places=1)
    self.assertAlmostEqual(result2[3], 0.5, places=1)


  def testPredictionMultipleCategories(self):
    """ Test the distribution of predictions.

    Here, we intend the classifier to learn the associations:
      [1,3,5] => bucketIdx 0 & 1
      [2,4,6] => bucketIdx 2 & 3

    The classifier should get the distribution almost right given enough
    repetitions and a small learning rate
    """
    c = Classifier( 0.001 )

    SDR1 = SDR(10);  SDR1.sparse = [1, 3, 5]
    SDR2 = SDR(10);  SDR2.sparse = [2, 4, 6]
    random.seed(42)
    for _ in range(5000):
      c.learn(pattern=SDR1, classification=[0, 1])
      c.learn(pattern=SDR2, classification=[2, 3])

    result1 = c.infer(pattern=SDR1)
    self.assertAlmostEqual(result1[0], 0.5, places=1)
    self.assertAlmostEqual(result1[1], 0.5, places=1)

    result2 = c.infer(pattern=SDR2)
    self.assertAlmostEqual(result2[2], 0.5, places=1)
    self.assertAlmostEqual(result2[3], 0.5, places=1)


  def testPredictionDistributionContinuousLearning(self):
    """ Test continuous learning

    First, we intend the classifier to learn the associations:
      SDR1    => bucketIdx 0 (30%)
              => bucketIdx 1 (30%)
              => bucketIdx 2 (40%)

      SDR2    => bucketIdx 1 (50%)
              => bucketIdx 3 (50%)

    After 20000 iterations, we change the association to
      SDR1    => bucketIdx 0 (30%)
              => bucketIdx 1 (20%)
              => bucketIdx 3 (40%)

      No further training for SDR2

    The classifier should adapt continuously and learn new associations for
    SDR1, but at the same time remember the old association for SDR2
    """
    c    = Classifier( 0.001 )
    SDR1 = SDR(10); SDR1.sparse = [1, 3, 5]
    SDR2 = SDR(10); SDR2.sparse = [2, 4, 6]

    random.seed(42)
    for _ in range(10000):
      randomNumber = random.random()
      if randomNumber < 0.3:
        bucketIdx = 0
      elif randomNumber < 0.6:
        bucketIdx = 1
      else:
        bucketIdx = 2
      c.learn( SDR1, bucketIdx )

      randomNumber = random.random()
      if randomNumber < 0.5:
        bucketIdx = 1
      else:
        bucketIdx = 3
      c.learn( SDR2, bucketIdx )

    result1 = c.infer( SDR1 )
    self.assertAlmostEqual(result1[0], 0.3, places=1)
    self.assertAlmostEqual(result1[1], 0.3, places=1)
    self.assertAlmostEqual(result1[2], 0.4, places=1)

    result2 = c.infer( SDR2 )
    self.assertAlmostEqual(result2[1], 0.5, places=1)
    self.assertAlmostEqual(result2[3], 0.5, places=1)

    for _ in range(20000):
      randomNumber = random.random()
      if randomNumber < 0.3:
        bucketIdx = 0
      elif randomNumber < 0.6:
        bucketIdx = 1
      else:
        bucketIdx = 3
      c.learn( SDR1, bucketIdx )

    result1new = c.infer( SDR1 )
    self.assertAlmostEqual(result1new[0], 0.3, places=1)
    self.assertAlmostEqual(result1new[1], 0.3, places=1)
    self.assertAlmostEqual(result1new[3], 0.4, places=1)

    result2new = c.infer( SDR2 )
    self.assertSequenceEqual(list(result2), list(result2new))


  def testMultiStepPredictions(self):
    """ Test multi-step predictions
    We train the 0-step and the 1-step classifiers simultaneously on
    data stream
    (SDR1, bucketIdx0)
    (SDR2, bucketIdx1)
    (SDR1, bucketIdx0)
    (SDR2, bucketIdx1)
    ...

    We intend the 0-step classifier to learn the associations:
      SDR1    => bucketIdx 0
      SDR2    => bucketIdx 1

    and the 1-step classifier to learn the associations
      SDR1    => bucketIdx 1
      SDR2    => bucketIdx 0
    """

    c = Predictor([0, 1], 1.0)

    SDR1 = SDR(10);  SDR1.sparse = [1, 3, 5]
    SDR2 = SDR(10);  SDR2.sparse = [2, 4, 6]
    recordNum = 0
    for _ in range(100):
      c.learn(recordNum, pattern=SDR1, classification=0)
      recordNum += 1

      c.learn(recordNum, pattern=SDR2, classification=1)
      recordNum += 1

    result1 = c.infer(recordNum, SDR1)
    result2 = c.infer(recordNum, SDR2)

    self.assertAlmostEqual(result1[0][0], 1.0, places=1)
    self.assertAlmostEqual(result1[0][1], 0.0, places=1)
    self.assertAlmostEqual(result2[0][0], 0.0, places=1)
    self.assertAlmostEqual(result2[0][1], 1.0, places=1)


  @unittest.skip("TODO: Pickle unimpemented!")
  def testSerialization(self):
    c = self._classifier([1], 1.0, 0.1, 0)
    c.compute(recordNum=0,
              patternNZ=[1, 5, 9],
              classification={"bucketIdx": 4, "actValue": 34.7},
              learn=True, infer=True)
    c.compute(recordNum=1,
              patternNZ=[0, 6, 9, 11],
              classification={"bucketIdx": 5, "actValue": 41.7},
              learn=True, infer=True)
    c.compute(recordNum=2,
              patternNZ=[6, 9],
              classification={"bucketIdx": 5, "actValue": 44.9},
              learn=True, infer=True)
    c.compute(recordNum=3,
              patternNZ=[1, 5, 9],
              classification={"bucketIdx": 4, "actValue": 42.9},
              learn=True, infer=True)
    serialized = pickle.dumps(c)
    c = pickle.loads(serialized)
    self.assertEqual(c.steps, [1])
    self.assertEqual(c.alpha, 1.0)
    self.assertEqual(c.actValueAlpha, 0.1)

    result = c.compute(recordNum=4,
              patternNZ=[1, 5, 9],
              classification={"bucketIdx": 4, "actValue": 34.7},
              learn=True, infer=True)
    self.assertSetEqual(set(result.keys()), set(("actualValues", 1)))
    self.assertAlmostEqual(result["actualValues"][4], 35.520000457763672,
                           places=5)
    self.assertAlmostEqual(result["actualValues"][5], 42.020000457763672,
                           places=5)
    self.assertEqual(len(result[1]), 6)
    self.assertAlmostEqual(result[1][0], 0.034234, places=5)
    self.assertAlmostEqual(result[1][1], 0.034234, places=5)
    self.assertAlmostEqual(result[1][2], 0.034234, places=5)
    self.assertAlmostEqual(result[1][3], 0.034234, places=5)
    self.assertAlmostEqual(result[1][4], 0.093058, places=5)
    self.assertAlmostEqual(result[1][5], 0.770004, places=5)


if __name__ == "__main__":
  unittest.main()
